src/diffusers/loaders/lora.py

```diff
@@ -1033,17 +1036,73 @@ def enable_lora_for_text_encoder(self, text_encoder: Optional["PreTrainedModel"]
     def set_adapters(
         self,
         adapter_names: Union[List[str], str],
-        adapter_weights: Optional[List[float]] = None,
+        adapter_weights: Optional[Union[List[float], List[Dict]]] = None,
     ):
+        adapter_names = [adapter_names] if isinstance(adapter_names, str) else adapter_names
+
+        number = (float, int)
+        has_2nd_text_encoder = hasattr(self, "text_encoder_2")
+
+        # Expand weights into a list, one entry per adapter
+        if adapter_weights is None or isinstance(adapter_weights, (number, dict)):
+            adapter_weights = [adapter_weights] * len(adapter_names)
+
+        if len(adapter_names) != len(adapter_weights):
+            raise ValueError(
+                f"Length of adapter names {len(adapter_names)} is not equal to the length of the weights {len(adapter_weights)}"
+            )
+
+        # Normalize into dicts
+        allowed_keys = ["text_encoder", "down", "mid", "up"]
+        if has_2nd_text_encoder:
+            allowed_keys.append("text_encoder_2")
+
+        def ensure_is_dict(weight):
+            if isinstance(weight, dict):
+                return weight
+            elif isinstance(weight, number):
+                return {key: weight for key in allowed_keys}
+            elif weight is None:
+                return {key: 1.0 for key in allowed_keys}
+            else:
+                raise RuntimeError(f"lora weight has wrong type {type(weight)}.")
+
+        adapter_weights = [ensure_is_dict(weight) for weight in adapter_weights]
+
+        for weights in adapter_weights:
+            for k in weights.keys():
+                if k not in allowed_keys:
+                    raise ValueError(
+                        f"Got invalid key '{k}' in lora weight dict. Allowed keys are 'text_encoder', 'text_encoder_2', 'down', 'mid', 'up'."
+                    )
```

Comment:
I wonder if it would help readability if all the new code above was factored out into a utility function like `def normalize_adapter_weight_argument(adapter_weights, has_2nd_text_encoder)` or so. Maybe diffusers maintainers have an opinion on that. Such a function would also be easier to unit-test.